vpr/src/place/timing_place.cpp

```diff
@@ -14,86 +14,93 @@
-
-static vtr::t_chunk f_timing_place_crit_ch;
-
-/******** prototypes ******************/
-static void alloc_crit(vtr::t_chunk* chunk_list_ptr);
-
-static void free_crit(vtr::t_chunk* chunk_list_ptr);
+//Use an incremental approach to updaing criticalities?
+#define INCR_UPDATE_CRITICALITIES
 
 /**************************************/
 
-/* Allocates space for the f_timing_place_crit data structure *
+/* Allocates space for the timing_place_crit_ data structure *
  * I chunk the data to save space on large problems.           */
-static void alloc_crit(vtr::t_chunk* chunk_list_ptr) {
-    auto& cluster_ctx = g_vpr_ctx.clustering();
-    float* tmp_ptr;
-
-    f_timing_place_crit.resize(cluster_ctx.clb_nlist.nets().size());
-
-    for (auto net_id : cluster_ctx.clb_nlist.nets()) {
-        tmp_ptr = (float*)vtr::chunk_malloc((cluster_ctx.clb_nlist.net_pins(net_id).size()) * sizeof(float), chunk_list_ptr);
-        f_timing_place_crit[net_id] = tmp_ptr;
-    }
+PlacerCriticalities::PlacerCriticalities(const ClusteredNetlist& clb_nlist, const ClusteredPinAtomPinsLookup& netlist_pin_lookup)
+    : clb_nlist_(clb_nlist)
+    , pin_lookup_(netlist_pin_lookup)
+    , timing_place_crit_(make_net_pins_matrix(clb_nlist_, std::numeric_limits<float>::quiet_NaN())) {
 }
 
 /**************************************/
-static void free_crit(vtr::t_chunk* chunk_list_ptr) {
-    vtr::free_chunk_memory(chunk_list_ptr);
-}
-
-/**************************************/
-void load_criticalities(SetupTimingInfo& timing_info, float crit_exponent, const ClusteredPinAtomPinsLookup& pin_lookup) {
-    /* Performs a 1-to-1 mapping from criticality to f_timing_place_crit.
+void PlacerCriticalities::update_criticalities(const SetupTimingInfo* timing_info, float crit_exponent) {
+    /* Performs a 1-to-1 mapping from criticality to timing_place_crit_.
      * For every pin on every net (or, equivalently, for every tedge ending
-     * in that pin), f_timing_place_crit = criticality^(criticality exponent) */
-
-    auto& cluster_ctx = g_vpr_ctx.clustering();
-    for (auto net_id : cluster_ctx.clb_nlist.nets()) {
-        if (cluster_ctx.clb_nlist.net_is_ignored(net_id))
-            continue;
-
-        for (auto clb_pin : cluster_ctx.clb_nlist.net_sinks(net_id)) {
-            int ipin = cluster_ctx.clb_nlist.pin_net_index(clb_pin);
-
-            float clb_pin_crit = calculate_clb_net_pin_criticality(timing_info, pin_lookup, clb_pin);
-
-            /* The placer likes a great deal of contrast between criticalities.
-             * Since path criticality varies much more than timing, we "sharpen" timing
-             * criticality by taking it to some power, crit_exponent (between 1 and 8 by default). */
-            f_timing_place_crit[net_id][ipin] = pow(clb_pin_crit, crit_exponent);
+     * in that pin), timing_place_crit_ = criticality^(criticality exponent) */
+
+    //Determine what pins need updating
+#ifdef INCR_UPDATE_CRITICALITIES
```

Comment:
Does this really need to be a #ifdef?  I'd be suprised if this being a runtime option is that expensive?

Reply:
I agree making this a run-time option is unlikely to be a performance issue. 

However these are fairly low-level configuration options which I do not think end-users should be concerned about/controlling. As a result, I don't see much point in exposing them at the command-line.

Reply:
I'm actually more objecting to the use of #ifdef's.  Consider that you have two ifdef's that control the same behavior in timing_place and slack_evaluation, I think the use of #ifdef's in this PR seems confusing and unclear to it's goal.

Reply:
The main motivations for having these options is to:

- Allow measurement of the performance characteristics of full vs incremental updates
- Support debugging by allowing various aspects of full vs incremental updates to be enabled/disabled

>  Consider that you have two ifdef's that control the same behavior in timing_place and slack_evaluation

To clarify, these defines actually all control different parts of the incremental update. There are multiple data structures/costs which are incrementally updated. Each has it's own define controlling whether it is fully or incrementally updated.

> A runtime bool, even if not exposed as a command line option, seems to make more sense to me.

Yes, this is reasonable and would avoid some potential code rot issues since all paths would always  be compiled.

Reply:
I've converted these define to bools.